src/mbgl/style/property_expression.cpp

```diff
@@ -1,14 +1,100 @@
+      zoomCurve(expression->has(Dependency::Zoom) ? expression::findZoomCurveChecked(*expression) : nullptr),
+      useIntegerZoom_(false),
+      isZoomConstant_(!expression->has(Dependency::Zoom)),
+      isFeatureConstant_(!expression->has(Dependency::Feature)),
+      isRuntimeConstant_(!expression->has(Dependency::Image)),
+      isGPUCapable_(checkGPUCapable(*expression, zoomCurve)) {
+    assert(isZoomConstant_ == expression::isZoomConstant(*expression));
+    assert(isFeatureConstant_ == expression::isFeatureConstant(*expression));
+    assert(isRuntimeConstant_ == expression::isRuntimeConstant(*expression));
+}
+
+PropertyExpressionBase::PropertyExpressionBase(PropertyExpressionBase&& other)
+    : expression(std::move(other.expression)),
+#if MLN_DRAWABLE_RENDERER
+      gpuExpression(std::move(other.gpuExpression)),
+#endif // MLN_DRAWABLE_RENDERER
+      zoomCurve(std::move(other.zoomCurve)),
+      useIntegerZoom_(other.useIntegerZoom_),
+      isZoomConstant_(other.isZoomConstant_),
+      isFeatureConstant_(other.isFeatureConstant_),
+      isRuntimeConstant_(other.isRuntimeConstant_),
+      isGPUCapable_(other.isGPUCapable_) {
+}
+
+PropertyExpressionBase::PropertyExpressionBase(const PropertyExpressionBase& other)
+    : expression(other.expression),
+#if MLN_DRAWABLE_RENDERER
+      gpuExpression(other.gpuExpression ? new gfx::GPUExpression(*(other.gpuExpression)) : nullptr),
+#endif // MLN_DRAWABLE_RENDERER
+      zoomCurve(other.zoomCurve),
+      useIntegerZoom_(other.useIntegerZoom_),
+      isZoomConstant_(other.isZoomConstant_),
+      isFeatureConstant_(other.isFeatureConstant_),
+      isRuntimeConstant_(other.isRuntimeConstant_),
+      isGPUCapable_(other.isGPUCapable_) {
+}
+
+PropertyExpressionBase& PropertyExpressionBase::operator=(PropertyExpressionBase&& other) {
+    expression = std::move(other.expression);
+#if MLN_DRAWABLE_RENDERER
+    gpuExpression = std::move(other.gpuExpression);
+#endif // MLN_DRAWABLE_RENDERER
+    zoomCurve = other.zoomCurve;
+    useIntegerZoom_ = other.useIntegerZoom_;
+    isZoomConstant_ = other.isZoomConstant_;
+    isFeatureConstant_ = other.isFeatureConstant_;
+    isRuntimeConstant_ = other.isRuntimeConstant_;
+    isGPUCapable_ = other.isGPUCapable_;
+    return *this;
+}
+
+PropertyExpressionBase& PropertyExpressionBase::operator=(const PropertyExpressionBase& other) {
+    expression = other.expression;
+#if MLN_DRAWABLE_RENDERER
+    gpuExpression.reset(other.gpuExpression ? new gfx::GPUExpression(*(other.gpuExpression)) : nullptr);
+#endif // MLN_DRAWABLE_RENDERER
+    zoomCurve = other.zoomCurve;
+    useIntegerZoom_ = other.useIntegerZoom_;
+    isZoomConstant_ = other.isZoomConstant_;
+    isFeatureConstant_ = other.isFeatureConstant_;
+    isRuntimeConstant_ = other.isRuntimeConstant_;
+    isGPUCapable_ = other.isGPUCapable_;
+    return *this;
+}
+
+const gfx::GPUExpression* PropertyExpressionBase::getGPUExpression([[maybe_unused]] bool intZoom) {
```

Comment:
Can you avoid using raw pointers?

Reply:
`shared_ptr` seems excessive here, as if it ever kept something alive it would be a bug, and a `reference_wrapper` or something doesn't seem to really add anything.

But mulling this over the weekend, I think it might work out better to create them in the `PropertyExpression` but storing/caching them elsewhere...

Reply:
I moved the ownership to the layer tweaker, passing them by reference to the property expression for potential updates, and that cleans up the semantics somewhat, though it makes the templates a little uglier.  Still needs some more cleanup, I think.